subsys/nrf_security/src/drivers/cracen/cracenpsa/src/key_derivation.c

```diff
@@ -0,0 +1,1182 @@
+						  uint8_t *output, size_t output_size,
+						  size_t *output_length)
+{
+	int sx_status;
+	psa_status_t psa_status;
+
+	char scratch_char_x[CRACEN_MAC_ECC_PRIVKEY_BYTES];
+	char scratch_char_y[CRACEN_MAC_ECC_PRIVKEY_BYTES];
+
+	struct sx_buf priv_key_buff = {.sz = priv_key_size, .bytes = (char *)priv_key};
+
+	sx_pk_affine_point scratch_pnt = {{.bytes = scratch_char_x, .sz = priv_key_size},
+					  {.bytes = scratch_char_y, .sz = priv_key_size}};
+	sx_pk_affine_point publ_key_pnt = {};
+
+	if (publ_key_size != cracen_ecc_wstr_expected_pub_key_bytes(priv_key_size)) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (publ_key[0] != SI_ECC_PUBKEY_UNCOMPRESSED) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	/* For Weierstrass curves in PSA we only use the X coordinate of the
+	 * point multiplication result.
+	 */
+	if (output_size < priv_key_size) {
+		return PSA_ERROR_BUFFER_TOO_SMALL;
+	}
+
+	publ_key_pnt.x.bytes = (char *)publ_key + 1;
+	publ_key_pnt.x.sz = priv_key_size;
+
+	publ_key_pnt.y.bytes = (char *)publ_key + priv_key_size + 1;
+	publ_key_pnt.y.sz = priv_key_size;
+
+	psa_status = cracen_ecc_check_public_key(curve, &publ_key_pnt);
+	if (psa_status != PSA_SUCCESS) {
+		return psa_status;
+	}
+
+	sx_status = sx_ecp_ptmult(curve, &priv_key_buff, &publ_key_pnt, &scratch_pnt);
+	if (sx_status == SX_OK) {
+		sx_pk_ecop2mem(&scratch_pnt.x, output, scratch_pnt.x.sz);
+		*output_length = scratch_pnt.x.sz;
+	}
+
+	safe_memzero(scratch_pnt.x.bytes, scratch_pnt.x.sz);
+	safe_memzero(scratch_pnt.y.bytes, scratch_pnt.y.sz);
+
+	return silex_statuscodes_to_psa(sx_status);
+}
+
+/**
+ * @brief Convert a byte array containing a scalar to sx_x25519_op.
+ *
+ * Loads a byte array (of size 32) and decodes it. This corresponds to the
+ * decodeScalar25519 function in RFC 7748.
+ */
+static struct sx_x25519_op decode_scalar_25519(const uint8_t *k)
+{
+	struct sx_x25519_op r;
+
+	memcpy(r.bytes, k, CRACEN_X25519_KEY_SIZE_BYTES);
+
+	r.bytes[0] &= 248;
+	r.bytes[31] &= 127;
+	r.bytes[31] |= 64;
+
+	return r;
+}
+
+/**
+ * @brief Convert a byte array containing a coordinate to sx_x25519_op.
+ *
+ * Loads a byte array (of size 32) and decodes it. This corresponds to the
+ * decodeUCoordinate (for 255 bits) function in RFC 7748.
+ */
+static struct sx_x25519_op decode_coordinate_25519(const uint8_t *k)
+{
+	struct sx_x25519_op r;
+
+	memcpy(r.bytes, k, CRACEN_X25519_KEY_SIZE_BYTES);
+
+	r.bytes[31] &= (1 << 7) - 1;
+
+	return r;
+}
+
+/**
+ * @brief Convert a byte array containing a scalar to sx_x448_op.
+ *
+ * Loads a byte array (of size 56) and decodes it. This corresponds to the
+ * decodeScalar448 function in RFC 7748.
+ */
+static struct sx_x448_op decode_scalar_448(const uint8_t *k)
+{
+	struct sx_x448_op r;
+
+	memcpy(r.bytes, k, CRACEN_X448_KEY_SIZE_BYTES);
+
+	r.bytes[0] &= 252;
+	r.bytes[55] |= 128;
+
+	return r;
+}
+
+static psa_status_t cracen_ecdh_montgmr_calc_secret(const struct sx_pk_ecurve *curve,
```

Comment:
Follow up task: Static function not documented

Reply:
I can't document this because I don't know what it does.